src/clims/services/substance.py

```diff
@@ -180,6 +180,18 @@ def project(self):
     def project(self, project):
         self._archetype.project = project._archetype
 
+    @property
+    def container_index(self):
+        # TODO: the container should be in a local cache!
+        # when looping samples in a container, the very same container is fetched over and
+        # over again
+        loc = self.location
+        container = self._app.containers.to_wrapper(loc.container)
+        container_index_class = container.IndexType
+        # Can we assume that container index class has only row and column, not z?
```

Comment:
No, the idea was that it should always have the three. I've got a suggestion of how to solve it which I'll commit to this PR. You can then rebase if you want to. I'll ping you when it's ready.

Reply:
Also, as @johandahlberg pointed out, this has the same role as the location property, but that's now returning these raw indexes. That one should be updated instead I think. I'll add that to my commit.

Reply:
And regarding the performance, I agree that we'll have to find a clean solution to that. Prefetching comes first to my mind - again.

Reply:
Ok